include/aws/crt/ImdsClient.h

```diff
@@ -0,0 +1,441 @@
+            };
+
+            /**
+             * https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
+             */
+            struct AWS_CRT_CPP_API IamProfileView
+            {
+                DateTime lastUpdated;
+                StringView instanceProfileArn;
+                StringView instanceProfileId;
+            };
+
+            /**
+             * A convenient class for you to persist data from IamProfileView, which has StringView members.
+             */
+            struct AWS_CRT_CPP_API IamProfile
+            {
+                IamProfile() {}
+                IamProfile(const IamProfileView &other)
+                    : lastUpdated(other.lastUpdated),
+                      instanceProfileArn(other.instanceProfileArn.data(), other.instanceProfileArn.size()),
+                      instanceProfileId(other.instanceProfileId.data(), other.instanceProfileId.size())
+                {
+                }
+
+                IamProfile &operator=(const IamProfileView &other)
+                {
+                    lastUpdated = other.lastUpdated;
+                    instanceProfileArn = String(other.instanceProfileArn.data(), other.instanceProfileArn.size());
+                    instanceProfileId = String(other.instanceProfileId.data(), other.instanceProfileId.size());
+                    return *this;
+                }
+
+                DateTime lastUpdated;
+                String instanceProfileArn;
+                String instanceProfileId;
+            };
+
+            /**
+             * Block of per-instance EC2-specific data
+             *
+             * https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html
+             */
+            struct AWS_CRT_CPP_API InstanceInfoView
+            {
+                /* an array of StringView */
+                Vector<StringView> marketplaceProductCodes;
+                StringView availabilityZone;
+                StringView privateIp;
+                StringView version;
+                StringView instanceId;
+                /* an array of StringView */
+                Vector<StringView> billingProducts;
+                StringView instanceType;
+                StringView accountId;
+                StringView imageId;
+                DateTime pendingTime;
+                StringView architecture;
+                StringView kernelId;
+                StringView ramdiskId;
+                StringView region;
+            };
+
+            /**
+             * A convenient class for you to persist data from InstanceInfoView, which has StringView members.
+             */
+            struct AWS_CRT_CPP_API InstanceInfo
+            {
+                InstanceInfo() {}
+                InstanceInfo(const InstanceInfoView &other)
+                    : availabilityZone(other.availabilityZone.data(), other.availabilityZone.size()),
```

Comment:
probably best to move these function implementations to the cpp file, to keep down the cost of including this header